panel/src/helpers/upload.js

```diff
@@ -94,6 +97,61 @@ export default async (file, params) => {
 			xhr.setRequestHeader(header, options.headers[header]);
 		}
 
+		// abort the XHR when abort signal is triggered
+		options.abort?.addEventListener("abort", () => {
+			xhr.abort();
+		});
+
 		xhr.send(data);
 	});
-};
+}
+
+/**
+ * Uploads a file in chunks
+ * @param {File} file - file to upload
+ * @param {Object} params - upload options (see `upload` method for details)
+ * @param {number} size - chunk size in bytes (default: 5 MB)
+ */
+export async function uploadAsChunks(file, params, size = 5242880) {
+	const parts = Math.ceil(file.size / size);
+	const id = random(4).toLowerCase();
+	let response;
+
+	for (let i = 0; i < parts; i++) {
+		// break if upload got aborted in the meantime
+		if (params.abort?.aborted) {
+			break;
+		}
+
+		// slice chunk at the right positions
+		const start = i * size;
+		const end = Math.min(start + size, file.size);
+		const chunk = parts > 1 ? file.slice(start, end, file.type) : file;
+
+		// when more than one part, add flag to
+		// recognize chunked upload and its last chunk
+		if (parts > 1) {
+			params.headers = {
+				...params.headers,
+				"Upload-Length": file.size,
+				"Upload-Offset": start,
+				"Upload-Id": id
+			};
+		}
+
+		response = await upload(chunk, {
+			...params,
+			// calculate the total progress based on chunk progress
+			progress: (xhr, chunk, percent) => {
+				const start = i * size;
```

Comment:
Is this line needed? We already set `start` to the same value in line 127 and this closure should be able to access that, no?

Reply:
Good catch